### PR TITLE
cephadm: refactor code to make use of DaemonIdentity type

### PR DESCRIPTION
**Pull request by phlogistonjohn**
Depends on #52178

Throughout cephadm a familiar pattern occurred: a function would have (at least) arguments of `fsid`, `daemon_type`, and `daemon_id`, not always in that order or exactly that naming but typically so.

In other words the code was generally expecting a triple of these values. This PR builds upon the work that added init_containers and the new `DaemonIdentity` type. DaemonIdentity is essentially this triple elevated to a type and with methods and attributes for commonly needed conversions (to a unit name for example).

I'll admit that this PR is pretty tedious and doesn't seem to do much. There's a lot of plumbing this new type in various places. You might be skeptical of the payoff of this work. In future work, we'll better make use of these types and also add additional helpful methods/attributes. But the scope of this PR is just to make use of DaemonIdentity, not make DaemonIdentity more useful. That said I do think there are benefits to correctness and type checking when you have a well defined type. DaemonIdentity uses the property decorator to implement `.fsid`, `.daemon_type`, and `.daemon_id` making the type quasi-read-only so that an error occurs if one tries to assign a new value to one of the properties.

Some functions make heavy use of all or one of `fsid`, `daemon_type`, and `daemon_id` in the body of the function. In many cases I simply assigned the DaemonIdentity property to a local variable (like `daemon_type = ident.daemon_type`) so as to avoid an unpleasant level of churn. But in cases where there were not a lot I used the DaemonIdentity value.property directly. When and where I chose to do this was somewhat subjective.

The code for adoption and list_units scares me. I only changed those where they needed to change. I otherwise left them alone because I don't know how to test them well enough yet.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [x] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), <s>opened tracker ticket</s>
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] Relies on existing tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by phlogistonjohn** [comment]
I'm planning to hold off on rebasing this until after PR #53052 is merged.

**Comment by phlogistonjohn** [comment]
I don't know what the docs build doesn't like about me lately. This time it seems cranky about the git commit hash...?  I don't think it's related to my changes though, so I feel I'm OK to move this out of draft finally.

**Comment by adk3798** [comment]
jenkins test docs

**Comment by phlogistonjohn** [comment]
Rebased, resolving a conflict with the tcmu runner wrapper script changes.

File: src/cephadm/cephadm.py
Comment (adk3798) [inline src/cephadm/cephadm.py:2473]: Is this just meant to be useful for unit testing purposes?

File: src/cephadm/cephadm.py
Comment (adk3798) [inline src/cephadm/cephadm.py:2621]: I'm a bit confused by this. If we want to make sure these mounts don't overwrite any of the other ones, wouldn't we want to do this first? So if any of the other mounts were on the same location they would overwrite the ones from here when the dict entry gets updated? It seems like the only one being set in this function checks first it isn't already in the mounts anyway so functionally it's the same, but the comment confuses me.

File: src/cephadm/cephadmlib/daemon_identity.py
Comment (adk3798) [inline src/cephadm/cephadmlib/daemon_identity.py:22]: Can I ask for the purpose of this? The types aren't marked optional so from a mypy POV it should already be making sure they're actually set and when actually running for a real cluster I wouldn't think just raising an assertion error would do us any good. Is this one of those commits to just check if we would ever hit this case when we do a teuthology run or something?

File: src/cephadm/cephadmlib/daemon_identity.py
Comment (adk3798) [inline src/cephadm/cephadmlib/daemon_identity.py:38]: NIT: The `<type>.<id>`  format is already used a lot as "daemon name" so I'd prefer that over "ceph name" unless we have some sort of conflict with something else called "daemon name" here. For example, a bunch of the type specific classes like the `CephIscsi` class have a `get_daemon_name` function that builds a name with this exact same format. It's also not limited to ceph daemons. The `Haproxy` and `Keepalived` class use the same convention.

**Review by adk3798 — APPROVED** [review]
LGTM. Also comes with added bonus of converting the type signatures of a lot of functions from py2 -> py3 format.

File: src/cephadm/cephadm.py
Comment (phlogistonjohn) [reply src/cephadm/cephadm.py:1895]: +1 on tracker

File: src/cephadm/cephadmlib/daemon_identity.py
Comment (phlogistonjohn) [reply src/cephadm/cephadmlib/daemon_identity.py:38]: Will do

File: src/cephadm/cephadm.py
Comment (phlogistonjohn) [reply src/cephadm/cephadm.py:2473]: Discussed in person.

File: src/cephadm/cephadm.py
Comment (phlogistonjohn) [reply src/cephadm/cephadm.py:2621]: I will simplify this comment.

File: src/cephadm/cephadm.py
Comment (phlogistonjohn) [reply src/cephadm/cephadm.py:2151]: It would be good for consistency but I think I'll skip it for effort.

File: src/cephadm/cephadm.py
Comment (phlogistonjohn) [reply src/cephadm/cephadm.py]: done

File: src/cephadm/cephadmlib/daemon_identity.py
Comment (phlogistonjohn) [reply src/cephadm/cephadmlib/daemon_identity.py]: done.

File: src/cephadm/cephadm.py
Comment (rkachach) [inline src/cephadm/cephadm.py:6248]: not sure why are we changing the behavior of the previous code, shouldn't this go in a separate PR?

**Review by rkachach — COMMENTED** [review]
Overall I like the idea of the change and using the new DaemonIdentity class, I think it makes the code more clear and easier to read :+1:. Not sure about if we really need the class  `DaemonSubIdentity`, cannot we just use the parent class?

File: src/cephadm/cephadm.py
Comment (phlogistonjohn) [reply src/cephadm/cephadm.py:6248]: Why do you think this is a behavior change?

**Comment by phlogistonjohn** [comment]
Regarding DaemonSubItentity - earlier versions had no DaemonSubIdentity and used an optional extra field on DaemonIdentity to identify sub-components of a Daemon. When I started building more code that used it that way I noticed it was easy to accidentally pass a DaemonIdentity without a subcomponent to code that required a subcomponent. (I was adding sidecars support)

That made me realize I should use the type system to make it harder to make mistakes like that. Thus all DaemonSubIdentity objects require/always-provide a sub-component. And they're used for id'ing init containers and sidecars - which will appear in a future PR.

File: src/cephadm/cephadm.py
Comment (phlogistonjohn) [reply src/cephadm/cephadm.py:1895]: https://tracker.ceph.com/issues/62816

**Comment by phlogistonjohn** [comment]
jenkins test make check

**Comment by adk3798** [comment]
https://pulpito.ceph.com/adking-2023-09-14_03:38:35-orch:cephadm-wip-adk5-testing-2023-09-13-1550-distro-default-smithi/

reruns of failed/dead jobs: https://pulpito.ceph.com/adking-2023-09-14_13:27:09-orch:cephadm-wip-adk5-testing-2023-09-13-1550-distro-default-smithi/

After reruns, 3 failures:

- 1 failure in mgr-nfs-upgrade, caused by a timeout when running `systemctl daemon-reload`. Don't have a tracker for this yet, but this happens in main baseline runs so unrelated to this PR
- 1 failure in the test_nfs task, this is known to not quite work currently
- 1 instance of https://tracker.ceph.com/issues/59704, known issue


Overall, nothing to block merging

**Pull request merged**
